deployment/managementgroup.go

```diff
@@ -517,16 +517,143 @@ func (mg *HierarchyManagementGroup) update(uniqueRoleDefinitions bool) error {
 	return nil
 }
 
+// ModifyPolicyAssignmentOption is a functional option for ModifyPolicyAssignment.
+type ModifyPolicyAssignmentOption func(*HierarchyManagementGroup, string) error
+
+// WithParameters sets the parameters for the policy assignment.
+func WithParameters(parameters map[string]*armpolicy.ParameterValuesValue) ModifyPolicyAssignmentOption {
+	return func(mg *HierarchyManagementGroup, name string) error {
+		pa := mg.policyAssignments[name]
+		if pa.Properties.Parameters == nil && len(parameters) > 0 {
+			pa.Properties.Parameters = make(
+				map[string]*armpolicy.ParameterValuesValue,
+				len(parameters),
+			)
+		}
+
+		for k, v := range parameters {
+			// Only add parameter if it exists in the referenced policy definition.
+			ref, policyDefinitionVersion, err := pa.ReferencedPolicyDefinitionResourceIDAndVersion()
+			if err != nil {
+				return fmt.Errorf(
+					"HierarchyManagementGroup.ModifyPolicyAssignment: "+
+						"error getting referenced policy definition resource id for policy assignment %s: %w",
+					name,
+					err,
+				)
+			}
+
+			if !mg.hierarchy.alzlib.AssignmentReferencedDefinitionHasParameter(ref, policyDefinitionVersion, k) {
+				return fmt.Errorf(
+					"HierarchyManagementGroup.ModifyPolicyAssignment: "+
+						"parameter `%s` not found in referenced %s `%s` for policy assignment `%s`",
+					k,
+					ref.ResourceType.Type,
+					ref.Name,
+					name,
+				)
+			}
+
+			pa.Properties.Parameters[k] = v
+		}
+
+		return nil
+	}
+}
+
+// WithEnforcementMode sets the enforcement mode for the policy assignment.
+func WithEnforcementMode(enforcementMode *armpolicy.EnforcementMode) ModifyPolicyAssignmentOption {
+	return func(mg *HierarchyManagementGroup, name string) error {
+		if enforcementMode != nil {
+			mg.policyAssignments[name].Properties.EnforcementMode = enforcementMode
+		}
+
+		return nil
+	}
+}
+
+// WithNonComplianceMessages sets the non-compliance messages for the policy assignment.
+func WithNonComplianceMessages(nonComplianceMessages []*armpolicy.NonComplianceMessage) ModifyPolicyAssignmentOption {
```

Comment:
@matt-FFFFFF Is the plan to add the more complex logic in the provider?

Reply:
see #228

I want to keep all logic in here just in case alzlib is used for anything else in future.

There will be some options attached to the hierarchy to control default non-compliance message behaviour. When the management group is created in memory the default non-compliance messages will be processed (1). After which and assignment-specific modifications can be applied (2).

Both of the above (1 & 2) operations will use the ModifyPolicyAssignments API that has been refactored in this PR